Zend/zend_generators.c

```diff
@@ -354,10 +353,19 @@ static HashTable *zend_generator_get_gc(zend_object *object, zval **table, int *
 	zend_execute_data *call = NULL;
 
 	if (!execute_data) {
-		/* If the generator has been closed, it can only hold on to three values: The value, key
-		 * and retval. These three zvals are stored sequentially starting at &generator->value. */
-		*table = &generator->value;
-		*n = 3;
+		if (UNEXPECTED(generator->func->common.fn_flags & ZEND_ACC_CLOSURE)) {
+			zend_get_gc_buffer *gc_buffer = zend_get_gc_buffer_create();
+			zend_get_gc_buffer_add_zval(gc_buffer, &generator->value);
+			zend_get_gc_buffer_add_zval(gc_buffer, &generator->key);
+			zend_get_gc_buffer_add_zval(gc_buffer, &generator->retval);
+			zend_get_gc_buffer_add_obj(gc_buffer, ZEND_CLOSURE_OBJECT(generator->func));
```

Comment:
Do we have a test for this? If the generator is not closed, how can it create a cycle without value/retval?

Reply:
Oh, by the closure capturing values I guess? Would be good to test this, unless I missed it.

Reply:
Without adding the closure object, these tests fail:

```
=====================================================================
FAILED TEST SUMMARY
---------------------------------------------------------------------
Collection of some cycles on unfinished generators [Zend/tests/bug69989_2.phpt]
A generator iterator wrapper involved in a cycle should not leak [Zend/tests/generators/iterator_wrapper_leak.phpt]
ReflectionGenerator while being currently executed [ext/reflection/tests/ReflectionGenerator_in_Generator.phpt]
=====================================================================
```

- https://github.com/php/php-src/blob/master/Zend/tests/bug69989_2.phpt
- https://github.com/php/php-src/blob/master/Zend/tests/generators/iterator_wrapper_leak.phpt
- https://github.com/php/php-src/blob/master/ext/reflection/tests/ReflectionGenerator_in_Generator.phpt

Reply:
This GC related change looks specious.
I think it stay the object alive before the next GC run (I may be wrong).
It would be good to add a test proving that the closure object is released together with generator.

Reply:
@dstogov I'm releasing the closure object in `zend_generator_free_storage()`. When I comment that out, then various tests fail due to memory leaks, for example [Zend/tests/arrow_functions/008.phpt](https://github.com/php/php-src/blob/master/Zend/tests/arrow_functions/008.phpt).

I've added an explicit test Zend/tests/generators/generator_closure_collection.phpt: b735b1b6a04ebb59fc01c0d7309110fd8fbf9d84